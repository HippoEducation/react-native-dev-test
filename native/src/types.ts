export type Post = {
  id: string;
  title: string;
  body: string;
  publishedAt: Date;
  author: {
    id: string;
    name: string;
  };
};

export type Posts = Post[];