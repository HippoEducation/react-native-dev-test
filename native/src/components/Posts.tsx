import React, { useState } from 'react';
import { StackNavigationProp } from '@react-navigation/stack';
import {
  StyleSheet,
  Text,
  View,
  ActivityIndicator,
} from 'react-native';
import useGetLatinPostService from '../services/useGetPostService';
import PostList from './PostList';
import { Post } from '../types/Post';
import { RootStackParamList } from '../types/RootStackParamList';

type PostsScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'Home'
>;

type Props = {
  navigation?: PostsScreenNavigationProp;
};

const Posts: React.FC<Props> = ({navigation}) => {
  const [posts, setPosts] = useState<Post[]>([]);
  
  const service = useGetLatinPostService(setPosts);

  const filterByAuthor = (name: string): void => setPosts(posts.filter((post) => post.author.name == name));

  const navigateToFullPost = (title: string, body: string): void => {
    navigation?.navigate('FullPost', { title: title, body: body });
  };

  return (
    <View style={styles.container}>
      {service.status === 'loading' && <ActivityIndicator />}
      {service.status === 'loaded' && <PostList 
        postList={posts}
        pressAuthorHandler={name => {filterByAuthor(name)}}
        pressPostHandler={(title, body) => navigateToFullPost(title, body)} 
      />}
      {service.status === 'error' && <Text>Oops, I did it again.</Text>}
    </View>
  );
};

export default Posts;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#000',
  },
});
