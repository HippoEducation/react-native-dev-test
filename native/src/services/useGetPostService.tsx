import { useEffect, useState } from 'react';
import { Post } from '../types/Post';
import { Service } from '../types/Service';

export interface Posts {
  posts: Post[];
}

const useGetPostService = () => {
  const [result, setResult] = useState<Service<Posts>>({
    status: 'loading',
  });

  useEffect(() => {
    fetch('http://10.0.0.122:4000/posts')
      .then((response) => response.json())
      .then((response) => setResult({ status: 'loaded', payload: response }))
      .catch((error) => setResult({ status: 'error', error }));
  }, []);

  return result;
};

export default useGetPostService;
