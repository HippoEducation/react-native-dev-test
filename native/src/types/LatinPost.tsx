export interface Author {
  id: string;
  name: string;
}

export interface LatinPost {
  author: Author;
  body: string;
  id: string;
  publishedAt: Date;
  title: string;
}
