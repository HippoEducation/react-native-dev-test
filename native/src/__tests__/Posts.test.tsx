import React from 'react';
import renderer, { create, act } from 'react-test-renderer';
import { mocked } from 'ts-jest/utils';
import 'isomorphic-fetch';
import useGetPostService from '../services/useGetPostService';
import { Service } from '../types/Service';
import { Post } from '../types/Post';
import Posts from '../components/Posts';

jest.mock('../services/useGetPostService');
const mockedService = mocked(useGetPostService, true);

describe('<Posts /> snapshots', () => {
  //I'd really like to learn how to successfully fire events to capture state change && navigation in snapshots
  it('displays ActivityIndicator while service status is still loading', async () => {
    const expectedResult: Service<Post[]> = {
      status: 'loading',
    };
    mockedService.mockReturnValueOnce(expectedResult);

    const tree = renderer.create(<Posts />).toJSON();
    await act(async () => {
      expect(tree).toMatchSnapshot();
    });
  });

  it('displays error message for error service status', async () => {
    const expectedResult: Service<Post[]> = {
      status: 'error',
      error: new Error(),
    };
    mockedService.mockReturnValueOnce(expectedResult);

    const tree = renderer.create(<Posts />).toJSON();
    await act(async () => {
      expect(tree).toMatchSnapshot();
    });
  });

  it('displays post list when results have loaded', async () => {
    const expectedResult: Service<Post[]> = {
      status: 'loaded',
      payload: [],
    };
    mockedService.mockReturnValueOnce(expectedResult);

    const tree = renderer.create(<Posts />).toJSON();
    await act(async () => {
      expect(tree).toMatchSnapshot();
    });
  });
});

it('renders ActivityIndicator component while loading', () => {
  const expectedResult: Service<Post[]> = {
    status: 'loading',
  };
  mockedService.mockReturnValueOnce(expectedResult);

  const tree = renderer.create(<Posts />).toJSON().children[0];

  expect(tree.type).toBe('ActivityIndicator');
});

it('displays error message when loaded with error', () => {
  const expectedResult: Service<Post[]> = {
    status: 'error',
    error: new Error(),
  };
  mockedService.mockReturnValueOnce(expectedResult);

  const tree = renderer.create(<Posts />).toJSON().children[0];

  expect(tree.type).not.toBe('ActivityIndicator');
  expect(tree.children[0]).toBe('Oops, I did it again.');
});
