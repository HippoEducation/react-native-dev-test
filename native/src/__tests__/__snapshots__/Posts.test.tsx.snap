// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Posts /> snapshots displays ActivityIndicator while service status is still loading 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;

exports[`<Posts /> snapshots displays error message for error service status 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Oops, I did it again.
  </Text>
</View>
`;

exports[`<Posts /> snapshots displays post list when results have loaded 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000",
      "flex": 1,
      "justifyContent": "center",
    }
  }
/>
`;
